src/server/services/etcd_meta_service.cc

```diff
@@ -52,37 +54,47 @@ void EtcdWatchHandler::operator()(etcd::Response const& resp) {
     head_rev = resp.events().back().kv().modified_index();
   }
 
-  std::vector<IMetaService::op_t> ops;
-  ops.reserve(resp.events().size());
+  std::lock_guard<std::mutex> scope_lock(this->session_add_mutex_);
+
+  std::map<std::string, std::vector<IMetaService::op_t>> ops;
   for (auto const& event : resp.events()) {
     std::string const& key = event.kv().key();
-    if (!boost::algorithm::starts_with(key, prefix_ + "/")) {
-      // ignore garbage values
-      continue;
-    }
-    if (!filter_prefix_.empty() &&
-        boost::algorithm::starts_with(key, filter_prefix_)) {
-      // FIXME: for simplicity, we don't care the instance-lock related keys.
-      continue;
-    }
-    IMetaService::op_t op;
-    std::string op_key = boost::algorithm::erase_head_copy(key, prefix_.size());
-    switch (event.event_type()) {
-    case etcd::Event::EventType::PUT: {
-      auto op = IMetaService::op_t::Put(op_key, event.kv().as_string(),
-                                        event.kv().modified_index());
-      ops.emplace_back(op);
-      break;
-    }
-    case etcd::Event::EventType::DELETE_: {
-      auto op = IMetaService::op_t::Del(op_key, event.kv().modified_index());
-      ops.emplace_back(op);
-      break;
-    }
-    default: {
-      // invalid event type.
-      break;
-    }
+    auto it = sessions_.begin();
+    while (it != sessions_.end()) {
+      if (!(*it)->meta_service_ptr->stopped()) {
+        if (boost::algorithm::boyer_moore_search<std::string>(key,
+                                                              "/meta_sync_lock")
+                .first != key.end()) {
+          break;
+        }
+        if (boost::algorithm::starts_with(key, (*it)->prefix + "/")) {
+          IMetaService::op_t op;
+          std::string op_key =
+              boost::algorithm::erase_head_copy(key, (*it)->prefix.size());
+          switch (event.event_type()) {
+          case etcd::Event::EventType::PUT: {
+            auto op = IMetaService::op_t::Put(op_key, event.kv().as_string(),
+                                              event.kv().modified_index());
+            ops[(*it)->sessionID].emplace_back(op);
+            break;
+          }
+          case etcd::Event::EventType::DELETE_: {
+            auto op =
+                IMetaService::op_t::Del(op_key, event.kv().modified_index());
+            ops[(*it)->sessionID].emplace_back(op);
+            break;
+          }
+          default: {
+            // invalid event type.
+            break;
+          }
+          }
+          break;
+        }
+        ++it;
+      } else {
+        it = sessions_.erase(it);
+      }
```

Comment:
I suggest you first split and group events to sessions, then dispatch events to each session, e.g.,

```
std::map<std::string /*prefix*/, std::vector<etcd::Event>>  event_groups;

for (auto const &event: resp.events()) {
    std::string const& key = event.kv().key();
    if (key.starts_with("/s")) { // it should belong to some session, and you know the length of session prefix
        event_groups[key.substr(1, 8 /* maybe */)].emplace_back(event);
    }
}

// then dispatch events to sessions

for (auto const &group: event_groups) {
    auto const &session = sessions_[group.first];
    for (auto const &event: group.second) {
         // processing the events, as existing logic
    }
}

```